### PR TITLE
Add support for updating JIRA Custom Fields

### PR DESCRIPTION
**Pull request by tequillaz**
Add JiraIssueFieldUpdater class for update issue custom field, 

Class is support Jenkins Pipeline Plugin. Pipeline example:
step([$class: 'JiraIssueFieldUpdater', field_id: '10100', field_value: '#1020', issueSelector: [$class: 'ExplicitIssueSelector', issueKeys: 'PI-11, PI-12, PI-13']])

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/jira-plugin/104)

### CODE REVIEW

File: src/main/java/hudson/plugins/jira/JiraIssueField.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package hudson.plugins.jira;
+
+import java.lang.String;
+
+public class JiraIssueField implements Comparable<JiraIssueField> {
+	
+	private final String field_id;
```
Comment (artkoshelev) [inline src/main/java/hudson/plugins/jira/JiraIssueField.java]: something wrong with indentation here and below

Comment (tequillaz) [inline src/main/java/hudson/plugins/jira/JiraIssueField.java]: looks like tab space. will fix this later

File: src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    public String field_value;
+
+    public String getFieldValue() {
+        return this.field_value;
+    }
+
+    @DataBoundSetter
+    public void setFieldValue(String field_value) {
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: please use consistent camelCase variable naming

File: src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    @DataBoundConstructor
+    public JiraIssueFieldUpdater(AbstractIssueSelector issueSelector, String field_id, String field_value) {
+        this.issueSelector = issueSelector;
+        this.field_id = field_id;
+        this.field_value = field_value;
+    }
+
+    public String prepare_field_id(String field_id) {
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: camelCasing

File: src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+        List<JiraIssueField> fields = Lists.newArrayList();
+        fields.add(new JiraIssueField(prepare_field_id(field_id), field_value));
+
+        for (String issue : issues) {
+            submitFeilds(session, issue, fields, logger);
+        }
+    }
+
+    public void submitFeilds(JiraSession session, String issueId, List<JiraIssueField> fields, PrintStream logger) {
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: typo in method name

File: src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            session.addFields(issueId, fields);
+        } catch (RestClientException e) {
+
+            if (e.getStatusCode().or(0).equals(404)) {
+                logger.println(issueId + " - JIRA issue not found. Dropping comment from update queue.");
+            }
+
+            if (e.getStatusCode().or(0).equals(403)) {
+                logger.println(issueId
+                        + " - Jenkins JIRA user does not have permissions to comment on this issue. Preserving comment for future update.");
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: * to  comment?
* how is this "preserving" for future update?

File: src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    public AbstractIssueSelector getIssueSelector() {
+        return this.issueSelector;
+    }
+
+    @DataBoundSetter
+    public void setIssueSelector(AbstractIssueSelector issueSelector) {
+        this.issueSelector = issueSelector;
+    }
+
+    public String field_id;
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: camelCasing

File: src/main/java/hudson/plugins/jira/JiraIssueField.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
+package hudson.plugins.jira;
+
+import java.lang.String;
+
+public class JiraIssueField implements Comparable<JiraIssueField> {
+
+    private final String field_id;
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraIssueField.java]: camelCasing

File: src/main/resources/hudson/plugins/jira/JiraIssueFieldUpdater/config.jelly [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <f:dropdownDescriptorSelector field="issueSelector" title="${%Issue selector}"/>
+    <f:entry field="field_id" title="Custom field id">
```
Comment (warden) [inline src/main/resources/hudson/plugins/jira/JiraIssueFieldUpdater/config.jelly]: id -> ID

File: src/main/resources/hudson/plugins/jira/JiraIssueFieldUpdater/help-field_id.html [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Issue custom field id. For example, 10100
```
Comment (warden) [inline src/main/resources/hudson/plugins/jira/JiraIssueFieldUpdater/help-field_id.html]: id -> ID

File: src/main/resources/hudson/plugins/jira/JiraIssueFieldUpdater/help.html [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<div>
+    JiraIssueFieldUpdater can be used to update custom jira issue field.
```
Comment (warden) [inline src/main/resources/hudson/plugins/jira/JiraIssueFieldUpdater/help.html]: jira -> JIRA

File: src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java (hunk 5 above)
Comment (tequillaz) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: Copyed this from Updater. I will remove "preserving comment..." from error message.

Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: in Updater the failed issues are preserved by:
https://github.com/jenkinsci/jira-plugin/blob/master/src/main/java/hudson/plugins/jira/Updater.java#L120

Here this logic is missing:
https://github.com/tequillaz/jira-plugin/blob/cd2d9db29b369df80f5ddc61a465765172042c3e/src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java#L131

File: src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        fields.add(new JiraIssueField(prepareFieldId(fieldId), fieldValue));
+
+        for (String issue : issues) {
+            submitFields(session, issue, fields, logger);
+        }
+    }
+
+    public void submitFields(JiraSession session, String issueId, List<JiraIssueField> fields, PrintStream logger) {
+        try {
+            session.addFields(issueId, fields);
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: Maybe you could implement JiraCarryOverAction similiar to how it is done in Updater? I'm pretty sure that's a useful thing when Jenkins will not have enough privileges... What do you think?

Comment (tequillaz) [inline src/main/java/hudson/plugins/jira/JiraIssueFieldUpdater.java]: If jenkins have no privileges we must notify about this jenkins administrator :) And he will solve the problem. Also i dont like way used by Updater with JiraCarryOverAction by next reasons
- it is not managable for end user
- similar behaviour must be implemented in every similar buildstep
- it is violate the single responsobility principe


I think best way is to create buildstep layer-wrapper to save and rerun not completed Jira* buildstep. We can move JiraCarryOverAction similar behaviour to this layer and this behaviour will work for all similar Jira* buildsteps.

**Comment by eskobaba** [comment]
I tried this and it will just give me an NPE

java.lang.NullPointerException
	at hudson.plugins.jira.JiraIssueFieldUpdater.prepareFieldId(JiraIssueFieldUpdater.java:85)
	at hudson.plugins.jira.JiraIssueFieldUpdater.perform(JiraIssueFieldUpdater.java:128)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:68)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:59)
	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1$1.call(AbstractSynchronousNonBlockingStepExecution.java:52)
	at hudson.security.ACL.impersonate(ACL.java:213)
	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1.run(AbstractSynchronousNonBlockingStepExecution.java:49)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

field_id exists, field_value is populated and issueKeys exists

**Comment by tequillaz** [comment]
@escobaba, can you explain how can i repeat this? look like fieldId is null string

**Comment by eskobaba** [comment]
@tequillaz actually i think i have another problem, it seems that anything other than 2.2.1 is giving me this:

javax.servlet.ServletException: java.lang.RuntimeException: java.util.concurrent.TimeoutException
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:796)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:233)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:135)
	at org.jenkinsci.plugins.ssegateway.Endpoint$SSEListenChannelFilter.doFilter(Endpoint.java:249)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at javax.servlet.FilterChain$doFilter.call(Unknown Source)
	at com.ceilfors.jenkins.plugins.jiratrigger.ExceptionLoggingFilter.doFilter(ExceptionLoggingFilter.groovy:22)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at hudson.plugins.greenballs.GreenBallFilter.doFilter(GreenBallFilter.java:59)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:201)
	at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:178)
	at net.bull.javamelody.PluginMonitoringFilter.doFilter(PluginMonitoringFilter.java:85)
	at org.jvnet.hudson.plugins.monitoring.HudsonMonitoringFilter.doFilter(HudsonMonitoringFilter.java:104)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at jenkins.metrics.impl.MetricsFilter.doFilter(MetricsFilter.java:125)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:126)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:553)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:95)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: java.util.concurrent.TimeoutException
	at com.google.common.base.Throwables.propagate(Throwables.java:156)
	at com.atlassian.httpclient.apache.httpcomponents.DefaultHttpClient$3.apply(DefaultHttpClient.java:256)
	at com.atlassian.httpclient.apache.httpcomponents.DefaultHttpClient$3.apply(DefaultHttpClient.java:249)
	at com.atlassian.util.concurrent.Promises$Of$2.apply(Promises.java:276)
	at com.atlassian.util.concurrent.Promises$Of$2.apply(Promises.java:272)
	at com.atlassian.util.concurrent.Promises$2.onFailure(Promises.java:167)
	at com.google.common.util.concurrent.Futures$6.run(Futures.java:977)
	at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:253)
	at com.google.common.util.concurrent.ExecutionList$RunnableExecutorPair.execute(ExecutionList.java:149)
	at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:134)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:188)
	at com.google.common.util.concurrent.SettableFuture.setException(SettableFuture.java:68)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient$1$3.run(SettableFuturePromiseHttpPromiseAsyncClient.java:73)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient$ThreadLocalDelegateRunnable$1.run(SettableFuturePromiseHttpPromiseAsyncClient.java:197)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient.runInContext(SettableFuturePromiseHttpPromiseAsyncClient.java:90)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient$ThreadLocalDelegateRunnable.run(SettableFuturePromiseHttpPromiseAsyncClient.java:192)
	... 3 more
Caused by: java.util.concurrent.TimeoutException
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient$1.doCancelled(SettableFuturePromiseHttpPromiseAsyncClient.java:67)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient$ThreadLocalContextAwareFutureCallback$3.run(SettableFuturePromiseHttpPromiseAsyncClient.java:152)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient.runInContext(SettableFuturePromiseHttpPromiseAsyncClient.java:90)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient$ThreadLocalContextAwareFutureCallback.cancelled(SettableFuturePromiseHttpPromiseAsyncClient.java:147)
	at org.apache.http.impl.client.cache.CachingHttpAsyncClient$2.cancelled(CachingHttpAsyncClient.java:636)
	at org.apache.http.concurrent.BasicFuture.cancel(BasicFuture.java:150)
	at org.apache.http.impl.nio.client.DefaultResultCallback.cancelled(DefaultResultCallback.java:57)
	at org.apache.http.impl.nio.client.DefaultAsyncRequestDirector.cancel(DefaultAsyncRequestDirector.java:533)
	at org.apache.http.impl.nio.client.AbstractHttpAsyncClient.doExecute(AbstractHttpAsyncClient.java:471)
	at org.apache.http.impl.nio.client.AbstractHttpAsyncClient.access$000(AbstractHttpAsyncClient.java:101)
	at org.apache.http.impl.nio.client.AbstractHttpAsyncClient$1.run(AbstractHttpAsyncClient.java:485)

reverting to 2.2.1 makes it fine. 
buildt with mvn install, tried to merge it against 2.2.1.

Can you please advise on how to merge/build to get it working?

**Comment by tequillaz** [comment]
@eskobaba, i did nothing special: maven build + manual plugin install with jenkins web interface 

After rebase to 2.3 i got connection timeout error with JIRA. I found that changing httpcomponents version in ea4deb1fabe92cb89736d432cef6d8b36e3b19db broke it. So i changed it version back and it works for me.

**Comment by eskobaba** [comment]
@tequillaz So you reverted all httpcomponents back to 4.2.5 or? Please provide more details so i can try this at work today. Thanks

**Comment by tequillaz** [comment]
@eskobaba, yes i reverted it back to 4.2.5. Try to set httpcomponnents version to 4.2.5 and rebuild jenkins jira-plugin with it.

**Comment by eskobaba** [comment]
@tequillaz: Odd, i have Fast-forwarded master to origin/pr/104. With 4.2.5 alot of test fails because of new stuff in httpcomponents 4.3+. Tried building without tests but still same failure when connecting. Tried building with 4.3 but still same problems.

Tried httpcore / httpcore-nio 4.4.5 instead of 4.4.4 too.

Would you be able to provide me with your hpi so i can test this?

**Comment by tequillaz** [comment]
@eskobaba, you need to downgrade version of all http components(httpmime, httpclient, httpcore, httpcore-nio, fluent-hc) to 4.2.5 as it was before. You can take hpi file here https://cloud.mail.ru/public/4D4m/SPBkejykf

**Comment by warden** [comment]
@tequillaz @eskobaba can you rebase and test with current master?

**Pull request merged**